### PR TITLE
Stream connection

### LINKED ISSUE #98: Implement DataSource for Stream<T>
I'm planing to add a generic implementation of `DataSource` for `Stream`s in the same fashion it's done for slice. I'm using mongodb on m project, which returns a `Stream` of documents.

I think it's kind of wasteful to collect all stream results into a vec to then apply `query_operation` filter on this vec. I don't know about other database systems, but I think most of them are using stream also?

Does anyone have thoughts on this?

### PR DESCRIPTION
**Pull request by IcanDivideBy0**
fixes #98

### CODE REVIEW

**Comment by IcanDivideBy0** [comment]
@phated @sunli829 tell me if its ok on the main idea, I'll add documentation tomorrow I think

File: src/types/connection/stream.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+}
+
+use std::collections::VecDeque;
+struct State<E: Send + Sync> {
+    has_seen_before: bool,
+    has_prev_page: bool,
+    has_next_page: bool,
+    edges: VecDeque<(Cursor, EmptyEdgeFields, E)>,
+}
```
Comment (IcanDivideBy0) [inline src/types/connection/stream.rs:33]: Not really happy with this, but I can't get a better idea :/

File: tests/connection.rs [hunk 2]
```diff
@@ -0,0 +1,282 @@
+pub mod stream {
+    use async_graphql::*;
+    use serde_json::{json, Value::Null};
```
Comment (sunli829) [inline tests/connection.rs:3]:
```suggestion
    use serde_json::json;
```

File: tests/connection.rs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+                    "edges": edges,
+                },
+            })
+        );
+    }
+
+    #[async_std::test]
+    pub async fn test_connection_stream_none() {
+        do_test(
+            json!({ "after": Null, "before": Null, "first": Null, "last": Null }),
```
Comment (sunli829) [inline tests/connection.rs:88]:
```suggestion
            json!({ "after": null, "before": null, "first": null, "last": null }),
```

File: tests/connection.rs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            let stream = futures::stream::iter(vec![
+                "a".to_owned(),
+                "b".to_owned(),
+                "c".to_owned(),
+                "d".to_owned(),
+                "e".to_owned(),
+                "f".to_owned(),
+            ]);
+
+            let mut source = StreamDataSource::new(stream, |s: &String| {
```
Comment (phated) [inline tests/connection.rs:26]: Is this iterating the entire stream to gather all cursors? If so, it seems inefficient that we'd iterate it twice

Comment (IcanDivideBy0) [reply tests/connection.rs:26]: No, the stream is just mapped with this closure:
https://github.com/async-graphql/async-graphql/pull/100/files#diff-c6747b5500cb9b8d48c79e5d07e7daa9R18

File: src/types/connection/stream.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+{
+    pub fn new<S>(stream: S) -> Self
+    where
+        S: Stream<Item = (Cursor, E, T)> + Send + Sync + 'a,
+    {
+        StreamDataSource {
+            stream: Box::pin(stream),
+        }
+    }
+}
```
Comment (IcanDivideBy0) [inline src/types/connection/stream.rs:27]: I've removed the mapping function for cursor, `StreamDataSource` now takes a `Stream<Item = (Cursor, E, T)>`. This gives more flexibility to users and is easy enouth for anyone to handle it in their codebase. Also gives the ability to control the EdgeFieldsObj. (see test)

Comment (IcanDivideBy0) [reply src/types/connection/stream.rs:27]: I think we should adopt the same strategy for the slice implementation.

**Comment by IcanDivideBy0** [comment]
Just some more thoughts: maybe we should remove the slice implementation in favor of anything that implement iterator ?
Then web can convert iterator into stream easily and use the same code for both ?

File: src/types/connection/stream.rs (hunk 5 above)
Comment (sunli829) [reply src/types/connection/stream.rs:27]: I agree. 😁

Comment (IcanDivideBy0) [reply src/types/connection/stream.rs]: removing the closure, we can now have `impl From<Stream> for StreamDataSource`, which is nice :)

**Comment by sunli829** [comment]
I fixed the problem that the master branch could not be compiled.

**Comment by sunli829** [comment]
@IcanDivideBy0  Can it be merged? 😁

**Comment by IcanDivideBy0** [comment]
@sunli829 yep, just wait for tests!

The current DataSource impl for slice is kind of whacky :
- cursor should not be used a base64 slice indices because push_front in the slice shifts all cursors to the next node, hence nodes may have different cursor based on their position
- edges fields are locked to `EmptyEdgeFields`
I think we should remove it but it would introduce breaking change...

**Comment by IcanDivideBy0** [comment]
@sunli829 please don't forget this is breaking change PR when releasing since `query_operation` now takes `&mut self`

**Comment by IcanDivideBy0** [comment]
wait, I think we can remove the `Sync` requirement for `DataSource`, which will simplify some code...

**Comment by IcanDivideBy0** [comment]
That's neat !
We can now query directly over a `BoxStream` :
```rust
async fn stream_connection(
    &self,
    ctx: &Context<'_>,
    after: Option<Cursor>,
    before: Option<Cursor>,
    first: Option<i32>,
    last: Option<i32>,
) -> FieldResult<Connection<String>> {
    futures::stream::iter(vec!["a", "b", "c", "d", "e", "f"])
    .map(|node| {
        let cursor: Cursor = node.clone().into();
        (cursor, EmptyEdgeFields, node)
    })
    .boxed()
    .query(ctx, after, before, first, last)
    .await
}
```

\o/

**Comment by sunli829** [comment]
The slice datasource is just for testing, and I don't believe anyone would actually use it. 😀

**Comment by sunli829** [comment]
> That's neat !
> 
> We can now query directly over a `BoxStream` :
> 
> ```rust
> 
> async fn stream_connection(
> 
>     &self,
> 
>     ctx: &Context<'_>,
> 
>     after: Option<Cursor>,
> 
>     before: Option<Cursor>,
> 
>     first: Option<i32>,
> 
>     last: Option<i32>,
> 
> ) -> FieldResult<Connection<String>> {
> 
>     let mut stream = futures::stream::iter(vec!["a", "b", "c", "d", "e", "f"])
> 
>         .map(|node| {
> 
>             let cursor: Cursor = node.clone().into();
> 
>             (cursor, EmptyEdgeFields, node)
> 
>         })
> 
>         .boxed();
> 
>     
> 
>     stream.query(ctx, after, before, first, last).await
> 
> }
> 
> ```
> 
> 
> 
> \o/


Amazingly easy to use!😂

**Comment by sunli829** [comment]
When you think there is no problem, you can merge them yourself. 
Thank you very much!😁

**Comment by IcanDivideBy0** [comment]
One more improvement and we're good to go:
I think a loop would be more efficient since I can break out of it anytime even if the stream is not terminated.

**Pull request merged**